### PR TITLE
🐛 Fix truncated file extraction from tarball

### PR DESCRIPTION
**Pull request by laurentsimon**
* **Please check if the PR fulfills these requirements**
- [X] Tests for the changes have been added (for bug fixes / features)
- [X] PR title follows the guidelines defined in  https://github.com/ossf/scorecard/blob/main/CONTRIBUTING.md#pr-process

* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)
bug fix

* **What is the current behavior?** (You can also link to an open issue here)
tarball files were truncated if the call to `tr.Read` did not return all the bytes at once.

* **What is the new behavior (if this is a feature change)?**
fix the above.

* **Does this PR introduce a breaking change?** (What changes might users need to make in their application due to this PR?)
no

* **Other information**:

### CODE REVIEW

File: checks/checkforcontent.go [hunk 1]
```diff
@@ -150,22 +161,22 @@ func CheckFilesContent(checkName, shellPathFnPattern string,
 			continue
 		}
 
-		content := make([]byte, hdr.Size)
-		n, err := tr.Read(content)
-		if err != nil && err != io.EOF {
-			return checker.MakeRetryResult(checkName, err)
+		content, err := readEntireFile(tr, hdr)
+		if err != nil {
+			return checker.MakeFailResult(checkName, err)
```
Comment (inferno-chromium) [inline checks/checkforcontent.go:166]: Why fail, why not retry.

File: checks/checkforcontent.go [hunk 2]
```diff
@@ -150,22 +161,22 @@ func CheckFilesContent(checkName, shellPathFnPattern string,
...
 		// the number of bytes should be the same as number
 		// indicated in header, unless the file format supports
 		// sparse regions. Only USTAR format does not support
 		// spare regions -- see https://golang.org/pkg/archive/tar/
-		if b := headerSizeMatchesFileSize(hdr, n); !b {
+		if b := headerSizeMatchesFileSize(hdr, len(content)); !b {
 			return checker.MakeRetryResult(checkName, ErrReadFile)
 		}
 
 		// We truncate the file to remove trailing 0 (sparse format).
```
Comment (inferno-chromium) [inline checks/checkforcontent.go:178]: is this comment still valid ?

File: checks/frozen_deps_test.go [hunk 3]
```diff
@@ -90,12 +92,14 @@ func TestGithubWorkflowPinning(t *testing.T) {
...
+			r, err := validateGitHubActionWorkflow(tt.args.Filename, content, tt.args.Log.Logf)
 
 			if !errors.Is(err, tt.want.Error) ||
-				r != tt.want.Result {
-				t.Errorf("TestGithubWorkflowPinning:\"%v\": %v (%v,%v) want (%v, %v)",
-					tt.name, tt.args.Filename, r, err, tt.want.Result, tt.want.Error)
+				r != tt.want.Result ||
+				len(tt.args.Log.messages) != tt.want.NumberOfErrors {
+				t.Errorf("TestDockerfileScriptDownload:\"%v\": %v (%v,%v,%v) want (%v, %v, %v)\n%v",
+					tt.name, tt.args.Filename, r, err, len(tt.args.Log.messages), tt.want.Result, tt.want.Error, tt.want.NumberOfErrors,
```
Comment (inferno-chromium) [inline checks/frozen_deps_test.go:101]: nit: Do we need that complicated message, similar is better ?

**Review by inferno-chromium — APPROVED** [review]

File: checks/checkforcontent.go [hunk 4]
```diff
@@ -89,6 +90,16 @@ func getTarReader(in io.Reader) (*tar.Reader, error) {
 	return tr, nil
 }
 
+func readEntireFile(tr *tar.Reader, hdr *tar.Header) (content []byte, err error) {
```
Comment (azeemshaikh38) [inline checks/checkforcontent.go:93]: tar.Header is not being used. Consider removing.

**Review by azeemshaikh38 — APPROVED** [review]

File: checks/checkforcontent.go (hunk 1 above)
Comment (laurentsimon) [reply checks/checkforcontent.go:166]: good point. Mistake from me :-)

File: checks/checkforcontent.go (hunk 2 above)
Comment (laurentsimon) [reply checks/checkforcontent.go:178]: you have lynx eyes!

File: checks/checkforcontent.go (hunk 4 above)
Comment (laurentsimon) [reply checks/checkforcontent.go:93]: good catch!

File: checks/frozen_deps_test.go (hunk 3 above)
Comment (laurentsimon) [reply checks/frozen_deps_test.go:103]: this gives all information we need to fix the problem. It only triggers when there's a failure.

**Pull request merged**